geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/GroupBy.scala

```diff
@@ -0,0 +1,119 @@
+/***********************************************************************
+* Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.utils.stats
+
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+case class GroupBy[T](attribute: Int, exampleStat: Stat)(implicit ct: ClassTag[T]) extends Stat {
+
+  override type S = GroupBy[T]
+
+  private [stats] val groupedStats: mutable.HashMap[T, Stat] = mutable.HashMap[T, Stat]()
+
+  def size: Int = groupedStats.size
+  def get(key: T): Option[Stat] = groupedStats.get(key)
+  def getOrElse[U >: Stat](key: T, default: U = null): U = groupedStats.getOrElse(key, default)
+
+  /**
+    * Compute statistics based upon the given simple feature.
+    * This method will be called for every SimpleFeature a query returns.
+    *
+    * @param sf feature to evaluate
+    */
+  override def observe(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.getOrElseUpdate(key, exampleStat.newCopy).observe(sf)
+  }
+
+  /**
+    * Tries to remove the given simple feature from the compiled statistics.
+    * Note: may not be possible to un-observe a feature, in which case this method will
+    * have no effect.
+    *
+    * @param sf feature to un-evaluate
+    */
+  override def unobserve(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key).foreach( groupedStat => groupedStat.unobserve(sf) )
+  }
+
+  /**
+    * Add another stat to this stat. Avoids allocating another object.
+    *
+    * @param other the other stat to add
+    */
+  override def +=(other: GroupBy[T]): Unit = {
+    other.groupedStats.map { case (key, stat) =>
+      groupedStats.get(key) match {
+        case Some(groupedStat) => groupedStat += stat
+        case None              => groupedStats.put(key, stat.newCopy)
+      }
+    }
+  }
+
+  /**
+    * Combine two stats into a new stat
+    *
+    * @param other the other stat to add
+    */
+  override def +(other: GroupBy[T]): GroupBy[T] = {
+    val newGB = new GroupBy[T](attribute, exampleStat.newCopy)
```

Comment:
[minor] don't need to copy exampleStat here since it's only used to create new copies